File: java/2015/day05a/Program.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.regex.*;

public class Program {
    public static void usage() {
        System.err.println("usage: java -jar Program.jar <input file>");
        System.exit(1);
    }

    public static boolean prop1(String line) {
        int vowels = 0;
        Pattern vowelPattern = Pattern.compile("[aeiou]");
        Matcher matcher = vowelPattern.matcher(line);
        while ( matcher.find() ) {
            vowels++;
        }
        return vowels >= 3;
    }

    public static boolean prop2(String line) {
        Pattern p = Pattern.compile("(.)\\1");
        Matcher m = p.matcher(line);
        return m.find();
    }

    public static boolean prop3(String line) {
        return !line.contains("ab") && !line.contains("cd") &&
            !line.contains("pq") && !line.contains("xy");
    }

    public static int process(String filename) {
        int result = 0;
        try ( BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            String line;
            while ( (line = reader.readLine()) != null ) {
                if ( prop1(line) && prop2(line) && prop3(line) ) {
                    result++;
                }
            }
        } catch ( IOException e ) {
            System.err.println("Error reading file: " + e.getMessage());
            System.exit(1);
        }
        return result;
    }

    public static final void main(String[] args) {
        if ( args.length < 1 ) {
            usage();
        }
        String filename = args[0];
        int result = process(filename);
        System.out.println("result = " + result);
    }
}

File: kotlin/2015/day05a/Program.kt
import java.io.File

fun usage() {
    System.err.println("usage: java -jar Program.jar <input file>")
    kotlin.system.exitProcess(1)
}

fun prop1(str: String) : Boolean {
    return Regex("[aeiou]").findAll(str).count() >= 3
}

fun prop2(str: String) : Boolean {
    return Regex("(.)\\1").containsMatchIn(str);
}

fun prop3(str: String) : Boolean {
    return !str.contains("ab") && !str.contains("cd") && !str.contains("pq") && !str.contains("xy")
}

fun process(content: String) : Int {
    var count = 0

    for ( line in content.split("\n") ) {
        if ( !line.isEmpty() ) {
            if ( prop1(line) && prop2(line) && prop3(line) ) {
                count += 1
            }
        }
    }

    return count
}

fun main(args: Array<String>) {
    if ( args.isEmpty()) {
        usage()
    }
    val filename = args[0]
    try {
        val content = File(filename).readText()
        val result = process(content)
        println("result = $result")
    } catch ( e: Exception ) {
        System.err.println("error processing file: ${e.message}")
    }
}
